### ISSUE TITLE
Represent composite edges in graphs with independent geometry.

### ISSUE THREAD

**Issue by olson-sean-k** [olson-sean-k/plexus#24]
`MeshGraph`s only expose half-edges. There is no notion of a composite edge in the public API and half-edges have independent geometry. This can make working with edge geometry somewhat awkward, as a composite edge's geometry is the combination of the geometry of its two half-edges.

Consider providing some notion of composite edges with singular geometry. For example, half-edges could share a singular geometry by referencing a shared composite edge. This could be a more natural representation. Internally, this would require a `CompositeEdge` topology to provide this geometry for half-edges.

With no directly associated geometry, half-edges would no longer have any kind of mutable iterators (orphan views) and could only expose geometry by reading their associated composite edge. A `CompositeEdgeView` type may only need to support consistent storage, as such a type wouldn't be very useful or necessary when mutating the topology of a mesh.

**Comment by olson-sean-k** [comment]
Rather than removing geometry from half-edges, it would be enough to introduce composite edges and allow users to choose whether or not to use that geometry. The `Geometry` trait would become:

```rust
pub trait Geometry: Sized {
    type Vertex: Attribute;
    type Half: Attribute + Default;
    type Edge: Attribute + Default;
    type Face: Attribute + Default;
}
```

If an application needs to associate geometry with composite edges but not half-edges, then it can simply implement the trait as follows:

```rust
impl Geometry for ColoredEdges {
    type Vertex = Point3<R64>;
    type Half = ();
    type Edge = Color4<R64>;
    type Face = ();
}
```

**Comment by olson-sean-k** [comment]
The [half](https://github.com/olson-sean-k/plexus/compare/half) branch currently shows some progress and experimentation toward this goal. It also changes nomenclature: the term _half_ is used to refer to half-edges while the term _edge_ is typically used to refer to composite edges.

```rust
let edge = face.into_edge().into_opposite_edge();
```
becomes
```rust
let half = face.into_half().into_opposite_half();
```

**Comment by olson-sean-k** [comment]
I may revisit the nomenclature here. I'd prefer succinct names for graph entities, but I find the term _half_ used alone to be a tad odd. For example:

```rust
let half = face.into_half();
```

It's unclear that "half" refers to _half-edge_ here. This could be especially confusing to anyone unfamiliar with half-edge graphs.

Perhaps _arc_ and _edge_ would be better. I haven't seen this nomenclature used for half-edge graphs before, but I have seen the terms used to distinguish between directed and undirected adjacency. Both terms are terse and easy to distinguish. Some small examples:

```rust
let face = graph.faces().nth(0).unwrap();
let arcs = graph
    .edges()
    .map(|edge| edge.into_arc())
    .collect::<Vec<_>>();
let edge = face.into_arc().into_edge();
let edge = face
    .into_arc()
    .into_opposite_arc()
    .into_next_arc()
    .into_edge();
```

**Comment by olson-sean-k** [comment]
While experimenting with this on the [half](https://github.com/olson-sean-k/plexus/compare/half) branch, I've noticed that the relationship of edges to other topological structures is not as cleanly hierarchical as vertices, arcs, and faces.

Moreover, some operations that I was tempted to move out of arcs and into edges don't work so well. For example, I attempted to move `split` and `remove` into `EdgeView`. This is possible, but cannot yield consistent results, because the connectivity to arcs is arbitrary. When splitting or removing an arc, the output topology can be oriented in terms of that arc. That's very useful, because it allows for traversals and mutations without direct access to the source `MeshGraph`. See the `subdivide` example. It seems that edges will probably be glorified mules for geometry while arcs remain the primary interface for manipulating both arcs and edges.
